internal/store/plugin.go

```diff
@@ -70,47 +68,57 @@ func (store *Store) GetPlugins(pluginFilter *model.PluginFilter) ([]*model.Plugi
 	return plugins[start:end], nil
 }
 
+// getPlugins returns all the plugins that satisfies serverVersion.
+// Slice is sorted by plugin name, ascending.
 func (store *Store) getPlugins(serverVersion string) ([]*model.Plugin, error) {
 	var result []*model.Plugin
-
-	for _, plugin := range store.plugins {
-		p, err := getPlugin(serverVersion, plugin)
-		if err != nil && err != ErrNotFound {
-			return nil, errors.Wrapf(err, "failed to get plugin")
-		} else if err != ErrNotFound {
-			result = append(result, p)
+	plugins := map[string]*model.Plugin{}
+
+	for _, storePlugin := range store.plugins {
+		meetsMinServerVersion := true
+		if serverVersion != "" && storePlugin.Manifest.MinServerVersion != "" {
+			meets, err := storePlugin.Manifest.MeetMinServerVersion(serverVersion)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to check minServerVersion for manifest.Id %s", storePlugin.Manifest.Id)
+			}
+			meetsMinServerVersion = meets
 		}
-	}
 
-	return result, nil
-}
-
-// getPlugin gets the first plugin from the sorted pluginVersions slice that satisfies serverVersion.
-func getPlugin(serverVersion string, pluginVersions pluginVersions) (*model.Plugin, error) {
-	if len(pluginVersions) == 0 {
-		return nil, errors.New("plugins should not be empty")
-	}
+		if !meetsMinServerVersion {
+			continue
+		}
 
-	// Get the latest plugin if no server version is provided
-	if serverVersion == "" {
-		return pluginVersions[len(pluginVersions)-1], nil
-	}
+		if plugins[storePlugin.Manifest.Id] == nil {
+			plugins[storePlugin.Manifest.Id] = storePlugin
+			continue
+		}
 
-	for i := len(pluginVersions) - 1; i >= 0; i-- {
-		// Missing MinServerVersion means it's compatible with all servers
-		if pluginVersions[i].Manifest.MinServerVersion == "" {
-			return pluginVersions[i], nil
+		pluginVersion, err := semver.Parse(plugins[storePlugin.Manifest.Id].Manifest.Version)
```

Comment:
As names, `pluginVersion` and `storePluginVersion` are subtle, since both came from the store. Perhaps a `lastSeen` variant instead?